tests/kernel/device/src/main.c

```diff
@@ -291,7 +291,7 @@ void test_dummy_device_pm(void)
 {
 	const struct device *dev;
 	int busy, ret;
-	unsigned int device_power_state = 0;
+	enum pm_device_state device_power_state;
```

Comment:
`device_power_state` is not directly assigned here, so it must be initialised to some value.

```suggestion
	enum pm_device_state device_power_state = -1;
```

Reply:
As far as I understand, the first time it is checked, it will have a valid assigned value:

```c
	ret = pm_device_state_get(dev, &device_power_state);
	zassert_true((ret == 0),
			"Unable to get active state to device");
	zassert_true((device_power_state == PM_DEVICE_STATE_ACTIVE),
			"Error power status");
```

Reply:
`pm_device_state_get(dev, &device_power_state);` in itself does not constitute an assignment -- i.e. there is no guarantee that `pm_device_state_get` modify `device_power_state` as far as the compiler is concerned.

`device_power_state` here is a stack variable and hence its initial value is unknown, meaning if `pm_device_state_get` does not modify `device_power_state` for whatever reason, this can lead to unpredictable behaviour.

Normally, you will get warnings about these (we used to IIRC), but somehow we aren't. I wonder what happened in that regard ...

Reply:
I understand the reason, but it should be considered a programming error or undefined behavior if `device_power_state` is used without checking the return code of `pm_device_state_get`.

Reply:
> it should be considered a programming error

Yes, a programming error that can go unnoticed if `pm_control` implementation does not write to `device_power_state` and the stack region happens to contain a value that corresponds to `PM_DEVICE_STATE_ACTIVE`.

There is a good reason why the existing implementation does `unsigned int device_power_state = 0;` -- i.e. doing so gets rid of any potential unpredictable behaviours.

This is a fairly standard practice and I don't see any reason why this should be an exception.

Reply:
> There is a good reason why the existing implementation does `unsigned int device_power_state = 0;` -- i.e. doing so gets rid of any potential unpredictable behaviours.
> 

No: If the pm_control function did not assign any value, the test will still pass because `PM_DEVICE_STATE_ACTIVE = 0`. Using `-1` doesn't still guarantee that one day the enum will use this value in its number range. I don't think it's a nice solution to initialize the variable with a value out of the enum range. If using Clang with `-Wassign-enum` this test would generate a build warning for the `-1` (`integer constant not in range of enumerated type`).

Reply:
The ideal solution here would be to add something like `PM_DEVICE_STATE_INVALID`.

Reply:
@stephanosio would it be acceptable to initialize with any state != `PM_DEVICE_STATE_ACTIVE` ? That means:

1. We assign a valid enum range
2. If the control function doesn't do an assignment, the first assert checking the expected value will fail.

Reply:
I agree with @gmarul. The variable should be initialized in `pm_device_state_get`. This function should return an error in case it not set the state. We have similar code all around the project, assign it to a valid value does not make sense if the control function does not behave properly. The only way would be assign to an invalid value as @stephanosio suggested, but I don't think it is necessary, a device should never be in a invalid state and this API does not interact with unknown or external code. That's said, I don't see any problem in what @gmarull  did.